numpy/random/mtrand/mtrand.pyx

```diff
@@ -4901,10 +4901,13 @@ cdef class RandomState:
         """
         if isinstance(x, (int, long, np.integer)):
             arr = np.arange(x)
-        else:
-            arr = np.array(x)
-        self.shuffle(arr)
-        return arr
+            self.shuffle(arr)
+            return arr
+
+        arr = np.array(x)
+        idx = np.arange(arr.shape[0])
+        self.shuffle(idx)
+        return arr[idx]
```

Comment:
I'd perhaps guard this with `if x.size * x.dtype.itemsize > np.dtype(np.intp) * len(x)`

Ie, check that the size of an index element is less than the actual slice being moved around

Reply:
I added some guard against using this method for 1d arrays which is fast.  This check added a lot fo complexity and didn't seem to make much difference for a very small 2-d array